registries/common.go

```diff
@@ -53,21 +81,106 @@ func (u *URL) FromString(raw string) error {
 	}
 	u.protocol = rawURL.Scheme
 	u.host = rawURL.Host
+	u.interfaceName = strings.TrimPrefix(rawURL.Path, "/")
 	u.params = rawURL.Query()
 	return nil
 }
 
+func (u *URL) ToString() string {
+	paramsPart, _ := url.QueryUnescape(u.params.Encode())
+	raw := fmt.Sprintf("%s://%s/%s?%s", u.protocol, u.host, u.interfaceName, paramsPart)
+	return url.QueryEscape(raw)
+}
+
 func (u *URL) ToInstance() discovery.Instance {
 	weight := DefaultDubboServiceWeight
-	if weightStr := u.params.Get("weight"); weightStr != "" {
+	if weightStr := u.params.Get(dubboInternalWeightKey); weightStr != "" {
 		if weightParam, err := strconv.Atoi(weightStr); err == nil {
 			weight = weightParam
 		}
 	}
 	params := map[string]string{
-		DubboServiceProtocolKey: u.protocol,
-		DubboServiceGroupKey:    u.params.Get("group"),
-		DubboServiceVersionKey:  u.params.Get("version"),
+		DubboServiceGroupKey:   u.params.Get(dubboInternalGroupKey),
+		DubboServiceVersionKey: u.params.Get(dubboInternalVersionKey),
 	}
 	return discovery.NewInstance("tcp", u.host, weight, params)
 }
+
+func (u *URL) FromInfo(info *registry.Info) error {
+	u.protocol = DefaultProtocol
+	if err := u.checkAndSetHost(info.Addr.String()); err != nil {
+		return err
+	}
+	if err := u.filterAndSetParams(info.Tags); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (u *URL) GetRegistryServiceKey(registryGroup string) string {
+	return fmt.Sprintf(RegistryServicesKeyTemplate, registryGroup, u.interfaceName)
+}
+
+func (u *URL) checkAndSetHost(addr string) error {
+	host, port, err := net.SplitHostPort(addr)
+	if err != nil {
+		return fmt.Errorf("parse addr %s error: %s", addr, err)
+	}
+	if port == "" {
+		return fmt.Errorf("addr %s missing port", addr)
+	}
+	if host == "" || host == "::" {
+		ipv4, err := getLocalIPV4Address()
+		if err != nil {
+			return fmt.Errorf("get local ipv4 error, cause %s", err)
+		}
+		u.host = ipv4 + ":" + port
+	} else {
+		u.host = host + ":" + port
+	}
+	return nil
+}
+
+func (u *URL) filterAndSetParams(params map[string]string) error {
+	var missingInterfaceFlag bool
+	var err error
+	defer func() {
+		if missingInterfaceFlag {
+			err = errors.New("tags must contain DubboServiceInterfaceKey:<interfaceName> pair")
+		}
+	}()
+	if len(params) <= 0 {
+		missingInterfaceFlag = true
+		return err
+	}
+	finalParams := make(url.Values)
+	for key, val := range params {
+		if dubboKey, ok := outboundDubboRegistryKeysMapping[key]; ok {
+			finalParams.Set(dubboKey, val)
+		}
+		if key == DubboServiceInterfaceKey {
+			u.interfaceName = val
+			missingInterfaceFlag = true
```

Comment:
Why not return the error directly?

Reply:
And is it wrongly implemented? missingInterfaceFlag should be initialized to be true and if there's a key `DubboServiceInterfaceKey` then we set it to false?

Reply:
I would add unit tests to fix it.